packages/client/src/client.ts

```diff
@@ -0,0 +1,303 @@
+  }
+
+  /** Return a record of all present providers, and fetch their manifests. */
+  public static getAllProviderManifests(): Record<string, Promise<PenumbraManifest>> {
+    return getPenumbraGlobalUnsafe() ? getAllPenumbraManifests() : {};
+  }
+
+  /** Fetch manifest of a specific provider, or return `undefined` if the
+   * provider is not present. */
+  public static getManifest(providerOrigin: string): Promise<PenumbraManifest> | undefined {
+    return getPenumbraUnsafe(providerOrigin) && getPenumbraManifest(providerOrigin);
+  }
+
+  /** Return boolean connection state of a specific provider, or `undefined` if
+   * the provider is not present. */
+  public static isConnected(providerOrigin: string): boolean | undefined {
+    return getPenumbraUnsafe(providerOrigin)?.isConnected();
+  }
+
+  /** Return connection state enum of a specific provider, or `undefined` if the
+   * provider is not present. */
+  public static getState(providerOrigin: string): PenumbraState | undefined {
+    return getPenumbraUnsafe(providerOrigin)?.state();
+  }
+
+  /** Make a connection attempt to a specific provider, and return a client bound
+   * to that provider if successful. May reject with an enumerated
+   * `PenumbraRequestFailure`. */
+  public static async connect(
+    requireProvider: string,
+    transportOptions = { jsonOptions } as ChannelTransportOptions,
+  ): Promise<PenumbraClient> {
+    await assertProviderManifest(requireProvider);
+    const client = new PenumbraClient(requireProvider, transportOptions);
+    await client.connect();
+    return client;
+  }
+
+  /** Create client, bound to `requireProvider` if specified, but do not attempt
+   * to request a connection. */
+  public static create(
+    requireProvider?: string,
+    transportOptions = { jsonOptions } as ChannelTransportOptions,
+  ) {
+    return new PenumbraClient(requireProvider, transportOptions);
+  }
+
+  // instance features
+
+  private readonly serviceClients = new Map<ServiceType['typeName'], PromiseClient<ServiceType>>();
+  private readonly stateListeners = new Set<
+    (detail: PenumbraEventDetail<'penumbrastate'>) => void
+  >();
+
+  private connection?: {
+    port: Promise<MessagePort>;
+    transport: Transport;
+  };
+
+  private config?: {
+    origin: string;
+    provider: PenumbraProvider;
+    requestManifest: Promise<PenumbraManifest>;
+    parsedManifest?: PenumbraManifest;
+  };
+
+  private readonly providerEventListener: PenumbraEventListener = evt => {
+    if (this.config?.origin && isPenumbraStateEvent(evt, this.config.origin)) {
+      this.stateListeners.forEach(listener => listener(evt.detail));
+    }
+  };
+
+  /** Construct an instance representing connection to a specific provider,
+   * identified by the origin parameter, but take to specific action. */
+  private constructor(
+    /** A provider origin may be provided during client construction. */
+    providerOrigin?: string,
```

Comment:
suggestion: instead of listing arguments, make it one extendable arguments interface like:

```ts
interface PenumbraClientConstructorOptions {
  providerOrigin?: string,
  transportOptions?: Omit<ChannelTransportOptions, 'getPort'>,
}

private constructor(options: PenumbraClientConstructorOptions) { ... }
```

We don't know how in the future we are going to extend the constructor but we can prepare for it. I think providing object options should apply here and in all functions for consistency and extendability

Reply:
i kept the origin identifier as a top-level parameter, as i'm fairly confident it should always be necessary. the second parameter now contains an options object.